Stripe/STPRedirectClient.m

```diff
@@ -0,0 +1,157 @@
+           presentingViewController:(UIViewController *)presentingViewController
+                         completion:(STPRedirectAuthCompletionBlock)completion {
+    [self cleanupAndCompleteRedirectWithError:nil];
+
+    if (![self redirectAllowedForSource:source]) {
+        return NO;
+    }
+
+    self.completion = completion;
+
+    // TODO: STPSource should provide these as named typed NSURLs instead of dictionary lookups
+    NSURL *redirectURL = [NSURL URLWithString:source.redirect[@"url"]];
+    NSURL *returnURL = [NSURL URLWithString:source.redirect[@"return_url"]];
+
+    // TODO: validate return URL
+
+    _inProgressAuthSource = source;
+
+    [[STPURLCallbackHandler shared] registerListener:self
+                                              forURL:returnURL];
+
+    if (!self.configuration.alwaysOpenSafari
+        && [SFSafariViewController class] != nil) {
+        SFSafariViewController *safariVC = [[SFSafariViewController alloc] initWithURL:redirectURL];
+        safariVC.delegate = self;
+        self.presentedViewController = safariVC;
+        [presentingViewController presentViewController:safariVC
+                                               animated:YES
+                                             completion:nil];
+    }
+    else {
+        // TODO: call openURL
+        [[UIApplication sharedApplication] openURL:redirectURL];
+    }
+
+    return YES;
+}
+
+- (BOOL)redirectAllowedForSource:(STPSource *)source {
+    // TODO: verify that they are valid URLs and that return url is set up correctly
+    return ((source.flow == STPSourceFlowRedirect)
+            && source.redirect[@"url"] != nil
+            && source.redirect[@"return_url"] != nil);
+}
+
+- (void)cleanupAndCompleteRedirectWithError:(nullable NSError *)error {
+    STPSource *source = self.inProgressAuthSource;
+
+    [[STPURLCallbackHandler shared] unregisterListener:self
+                                                forURL:[NSURL URLWithString:source.redirect[@"return_url"]]];
+
+
+    if (self.presentedViewController) {
+        [self.presentedViewController dismissViewControllerAnimated:YES
+                                                         completion:^{
+                                                             if (self.completion) {
+                                                                 self.completion(source, error);
+                                                             }
+                                                             self.presentedViewController = nil;
+                                                         }];
+    }
+    else {
+        if (self.completion) {
+            self.completion(source, error);
+        }
+    }
+
+    _inProgressAuthSource = nil;
+    self.completion = nil;
+
+}
+
+- (BOOL)handleURLCallback:(NSURL *)url {
+    if (url == nil
+        || self.inProgressAuthSource == nil) {
+        [self cleanupAndCompleteRedirectWithError:nil];
```

Comment:
Yeah I am not 100% sure on all the errors yet. Also debating whether there should just be a "succeeded" BOOL or something, but not sure how to detect easily whether a redirect "completed" or if we're still waiting for confirmation. Will go through and fill in errors later once the shape of the thing is more fleshed out.